#330 - Enable deep copy

dmnc-grdnr opened the pull request:
This PR adds the deep copy functionality to the CAS object.

Tests for the function was added to test_cas.py

I have verified the functionality on different CAS files, but due to the complexity of the data structure, there might be something I have overlooked. 

Therefore feedback is highly appreciated!

dmnc-grdnr committed:
feature/330-enable-deep-copy

reckart committed:
Merge branch 'main' into feature/330-enable-deep-copy

dmnc-grdnr commented on cassis/cas.py:
```diff
@@ -832,6 +833,105 @@ def _copy(self) -> "Cas":
+                            referenced_lists[feature.name].append((fs.xmiID, referenced_list))
+                    elif ts.is_array(feature.rangeType):
+                        referenced_list = []
+                        for item in fs[feature.name].elements:
```
I disagree with this argument. However it does not matter, because lines 898-902 are actually dead code (connected to unused variable referenced_lists see comment above). These lines should be removed instead.

dmnc-grdnr commented:
I reviewed the Copilot suggestions agreed with most, but found that a few lines are actually obsolete and should be removed instead of modified. This is my first pull request and I am unsure how to make the changes without breaking things. Could you please point me to what to do next?

reckart reviewed (dismissed):
Thank you very much. To continue, it would be great if you could simply address the comments by updating the code and then committing the changes into this branch. Once the changes are in, the review will be repeated. Please do not use the "amend" function of git, just add another commit. Avoid force pushes, then you can't really break anything. Thanks!

dmnc-grdnr committed:
revised and updated code for deep copy; added random CAS tests

dmnc-grdnr commented on cassis/cas.py:
```diff
@@ -832,6 +833,119 @@ def _copy(self) -> "Cas":
+
+        # add feature structures to the appropriate views
+        feature_structures = sorted(all_copied_fs.values(), key=lambda f: f.xmiID, reverse=False)
+        for item in all_copied_fs.values():
```
Yes, but I would remove the unused variable instead.

dmnc-grdnr commented:
Hello @reckart,

it took me some time, but I have not lost interest :) 
I have read through all of the suggestions and commented wherever I was sure about the answer. The suggestions left uncommented (in my opinion) boil down to questions of defensive programming vs plausibility. What I mean is, should the code be able to copy a structurally damaged CAS or should it fail instead: 
Should it copy a feature structure that does not have a name attribute? 
Should it copy an FSArray, that contains None-elements or an FSArray which has an elements-attribute set to the value None? 
Should the code expect there to be feature structures that are not returned by cas._find_all_fs but somehow their xmIDs are referenced in other feature structures?

Don't get me wrong, I don't mind applying these safety measuers at all. It just seems that they might be superflous. Then again my analysis might superficial instead ;) 

(I'd be happy to have call about it too, if you like)

reckart commented:
@dmnc-grdnr  Best write me a mail.

reckart committed:
Merge branch 'main' into feature/330-enable-deep-copy

reckart committed:
Merge branch 'main' into feature/330-enable-deep-copy

* main:
  #328 - Remove annotations in a specified range
  #328 - Remove annotations in a specified range
  #328 - Remove annotations in a specified range
  #328 - Remove annotations in a specified range
  #328 - Remove annotations in a specified range
  #328 - Remove annotations in a specified range
  #328 - Remove annotations in a specified range
  #328 - Remove annotations in a specified range
  Issue #342: Improve cas_to_comparable_text
  #328 - Remove annotations in a specified range
  #328 - Remove annotations in a specified range

% Conflicts:
%	tests/test_cas.py

reckart committed:
Issue #330: Enable deep copy

- ensure original and copy are fully decoupled
- Preserve None entries in FSArray copies and avoid warnings for None non-primitive features.
- Avoid creating a document annotation when checking for the language of the source CAS
- Support FSList and FSArray in cloning

reckart committed:
Issue #330: Enable deep copy

- properly handle collection features that have None value

reckart committed:
Issue #330: Enable deep copy

- Avoid shared collection objects between original and clone

reckart committed:
Issue #330: Enable deep copy

- Fix handling of primitive lists

reckart committed:
Issue #330: Enable deep copy

- Fix handling of inlined vs. reusable arrays

reckart committed:
Issue #330: Enable deep copy

- Fix handling of inlined vs. reusable primitive arrays

reckart committed:
Issue #330: Enable deep copy

- Added more testing

reckart committed:
Issue #330: Enable deep copy

- Fix handling of inlined vs. reusable primitive arrays

reckart committed:
Issue #330: Enable deep copy

- Fix remembering which structures should be in the index and in which index and which should not be indexed

reckart committed:
Issue #330: Enable deep copy

- Remove print calls from tests

reckart committed:
Issue #330: Enable deep copy

- Better tracking of which FS belongs to which view(s)

reckart commented on cassis/cas.py:
```diff
@@ -937,6 +938,314 @@ def _copy(self) -> "Cas":
+                    referenced_view[member.xmiID] = view.sofa.sofaID
+
+        for fs in self._find_all_fs():
+            t = ts.get_type(fs.type.name)
```
YAGNI

reckart committed:
Issue #330: Enable deep copy

- No need to pass language and mime-type at CAS construction

reckart committed:
Issue #330: Enable deep copy

- Also properly duplicate sofaArray if set

reckart committed:
Issue #330: Enable deep copy

- Preserve active view

reckart committed:
Merge branch 'main' into feature/330-enable-deep-copy

* main:
  #336 - Introduce Annotation class
  #336 - Introduce Annotation class
  #336 - Introduce Annotation class
  #336 - Introduce Annotation class
  #336 - Introduce Annotation class
  Issue #349: Upgrade dependencies
  Issue #349: Upgrade dependencies
  #336 - Introduce Annotation class

% Conflicts:
%	cassis/cas.py
%	tests/test_cas.py
%	tests/test_json.py
%	tests/test_util.py

reckart committed:
#330 - Resolve merge conflicts and fix deep_copy regressions from main

- Resolve merge conflicts in cassis/cas.py imports and in test_cas.py / test_json.py / test_util.py (kept deep_copy tests + main's runtime-generated annotation test; dropped debug print statements).
- Fix view-membership regression in deep_copy: replace the now-deprecated view.get_all_annotations() / Cas.select_all() (which post-merge filter to true Annotation instances) with view.get_all_fs() / select_all_fs() so non-annotation indexed FS are restored to view indices in the copy.
- Update test_xmi.py / test_json.py callers from .select_all() to .select_all_fs() to match the tightened post-merge contract.
- Handle concrete primitive-list rangeTypes (Empty*/NonEmpty* Integer/Float/StringList) in deep_copy via _PRIMITIVE_LIST_BASE_TYPE map; previously these matched is_primitive_collection but neither is_array nor is_list, so the value was silently dropped on copy.
- Add an else: warnings.warn(...) at the inner primitive-collection branch so any genuinely unhandled shape surfaces instead of dropping data silently (per reckart's review feedback).
- Rename five same-scope reuses of `referenced_list` in deep_copy to distinct, descriptive names (standalone_fs_array_member_ids, array_feature_member_ids, array_member_ids, fs_array_member_ids, fs_list_member_ids).
- Add deep_copy edge-case tests: empty primitive array, None FSList feature, empty FSList feature, deep_copy of empty CAS, plus three regression tests for concrete primitive-list rangeType handling.

reckart committed:
#3411 - Resize sidebars using mouse

- Improved failure mode when trying to copy data to a target CAS missing the source type
